components/list_modal.jsx

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import {Modal} from 'react-bootstrap';
+import {FormattedMessage} from 'react-intl';
+
+import LoadingScreen from 'components/loading_screen.jsx';
+
+export const DEFAULT_NUM_PER_PAGE = 50;
+
+export default class ListModal extends React.PureComponent {
+    static propTypes = {
+
+        /**
+         * loadItems is a function that receives the params (pageNumber, searchTerm) and should return an object
+         * with the shape {items: [], totalCount: 0}.
+         * items: an array of objects that are passed to each renderRow function.
+         * totalCount: an integer representing the total number of items as displayed in the pagination text.
+         *
+         * Example:
+         *     const loadItems = async (pageNumber, searchTerm) => {
+         *         const {data} = await loadFromServer(searchTerm, pageNumber, PER_PAGE);
+         *         return {
+         *             items: data.users,
+         *             totalCount: data.total,
+         *         };
+         *     };
+         */
+        loadItems: PropTypes.func.isRequired,
+
+        /**
+         * renderRow is a function that receives the params (item, listModal) and should return JSX.
+         * item: an object as returned by each entry in the loadItems function's 'items' array.
+         * listModal: the instance of the ListModal component class.
+         *
+         * Example:
+         *     const renderRow = (item, listModal) => <div>{item.id}</div>;
+         */
+        renderRow: PropTypes.func.isRequired,
+
+        /**
+         * onHide (optional) a function to be invoked when the modal is closed.
+         */
+        onHide: PropTypes.func,
+
+        /**
+         * titleText (optional) a string to show at the top bar of the modal.
+         */
+        titleText: PropTypes.string,
+
+        /**
+         * searchPlaceholderText (optional) a string to show as a placeholder in the search input.
+         */
+        searchPlaceholderText: PropTypes.string,
+
+        /**
+         * titleBarButton (optional) an object representing a title button.
+         */
+        titleBarButton: PropTypes.shape({
```

Comment:
You are passing here an object that you build each time when you use the component, and at the same time you are defining the component as PureComponent, so, or you improve how you pass this object to be sure the object is not changing, or you use the Component.

Reply:
I agree with you in theory but in my testing the number of renders don't change whether it's the way it is now, whether it has a `shouldComponentUpdate` method as a `Component`, or whether it's a `PureComponent` with the `titleBarButtonText` and `titleBarButtonOnClick` params. I wonder if my test is broken.

Reply:
I wonder if the vDOM is handling it and avoiding the re-render but still causing extra cycles in the JS code.